### PR TITLE
#182461142 - USERS SHOULD BE ABLE TO FOLLOW EACH OTHER

### PR DESCRIPTION
**Pull request by ChrisOnsando**
#### What does this Pull Request do?
- [ ] Users are able to follow and unfollow each other
#### Description of Task to be completed?
- [ ] Users should be able to follow and unfollow each other
- [ ] Users should be able to see their followers and following
#### How should this be manually tested?
- [ ] http://127.0.0.1:8000/api/user/follow/
#### Any background context you want to provide?
- [ ] Users should follow and unfollow users they wish
- [ ] Users should be see all their followers and following
#### What are the relevant Pivotal Tracker story?
- [ ]  https://www.pivotaltracker.com/story/show/182461142
#### Screenshots (if appropriate)
Screenshot from 2022-07-28 13-47-51
Screenshot from 2022-07-29 10-56-29

#### Questions:
- [ ] None

### CODE REVIEW

File: app/user/models.py [hunk 1]
```diff
@@ -96,3 +102,39 @@ class Profile(models.Model):
 
     def __str__(self) -> str:
         return self.user.username
+
+
+class UserFollowing(TimeStampedModel):
+    id = models.UUIDField(
```
Comment (mainamuriuki274) [inline app/user/models.py:108]: I told you to remove this field

File: app/user/tests/test_views.py [hunk 2]
```diff
@@ -246,3 +247,83 @@ def test_invalid_user_id(self) -> None:
...
+        self.password = fake.password()
+        self.user_one = User.objects.create_user(
+            username=fake.name(), email=fake.email(), password=self.password
+        )
+        self.user_two = User.objects.create_user(
+            username=fake.name(), email=fake.email(), password=self.password
+        )
+        self.client = APIClient()
+
+    @property
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:263]: remove this decorator

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

File: app/user/models.py (hunk 1 above)
Comment (ChrisOnsando) [reply app/user/models.py:108]: Change made

File: app/user/tests/test_views.py [hunk 3]
```diff
@@ -246,3 +247,104 @@ def test_invalid_user_id(self) -> None:
...
+
+    def test_authorized_user_follow(self) -> None:
+        url = reverse("follow")
+        response = self.client.post(
+            url,
+            data={"follow": self.user_two.id},
+            format="json",
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:300]: also assert that the user is in the followers list

File: app/user/tests/test_views.py [hunk 4]
```diff
@@ -246,3 +247,104 @@ def test_invalid_user_id(self) -> None:
...
+            format="json",
+            **self.bearer_token,
+        )
+        url = reverse("unfollow", kwargs={"id": self.user_two.id})
+        response = self.client.delete(
+            url,
+            format="json",
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:316]: assert that the user is no longer in the followers list

File: app/user/tests/test_views.py [hunk 5]
```diff
@@ -246,3 +247,104 @@ def test_invalid_user_id(self) -> None:
...
+
+    def test_unauthorized_user_unfollow(self) -> None:
+        url = reverse("unfollow", kwargs={"id": self.user_one.id})
+        response = self.client.delete(url, format="json")
+        self.assertEqual(response.status_code, status.HTTP_401_UNAUTHORIZED)
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:321]: assert for the expected message also

File: app/user/tests/test_views.py [hunk 6]
```diff
@@ -246,3 +247,104 @@ def test_invalid_user_id(self) -> None:
...
+
+    def test_unauthorized_user_follow(self) -> None:
+        url = reverse("follow")
+        response = self.client.post(
+            url,
+            data={"follow": self.user_one.id},
+            format="json",
+            **self.bearer_token,
+        )
+        self.assertEqual(response.status_code, status.HTTP_201_CREATED)
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:331]: How is this one 201 and the user is unauthorized???? don't just write tests for the sake

File: app/user/tests/test_views.py [hunk 7]
```diff
@@ -246,3 +247,104 @@ def test_invalid_user_id(self) -> None:
         resp = self.client.post(reset_url)
         self.assertEqual(resp.status_code, 400)
         self.assertIn("This field is required", str(resp.data))  # type: ignore[attr-defined]
+
+
+class TestUserFollowingView(APITestCase):
+    def setUp(self) -> None:
+        self.password = fake.password()
+        self.user_one = User.objects.create_user(
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:255]: Check on this

File: app/user/tests/test_views.py [hunk 8]
```diff
@@ -246,3 +247,104 @@ def test_invalid_user_id(self) -> None:
...
+            username=fake.name(), email=fake.email(), password=self.password
+        )
+        self.user_two = User.objects.create_user(
```
Comment (mainamuriuki274) [inline app/user/tests/test_views.py:258]: Check on this

**Review by mainamuriuki274 — CHANGES_REQUESTED** [review]

**Comment by dalienst** [comment]
PR is ready to be merged

**Review by mainamuriuki274 — APPROVED** [review]

**Pull request merged**